WRITING-26459 Add v2 Migration Guide (#1766)
mongodb/mongo-go-driver: docs/migration-2.0.md

- `PoolCreated` to `ConnectionPoolCreated`
- `PoolReady` to `ConnectionPoolRead`
- `PoolCleared` to `ConnectionPoolCleared`
- `PoolClosedEvent` to `ConnectionPoolClosed`
- `GetStarted` to `ConnectionCheckOutStarted`
- `GetFailed` to `ConnectionCheckOutFailed`
- `GetSucceeded` to `ConnectionCheckedOut`
- `ConnectionReturned` to `ConnectionCheckedIn`

### CommandFailedEvent

`CommandFailedEvent.Failure` has been converted from a string type to an error type to convey additional error information and to match the type of other `*.Failure` fields, like `ServerHeartbeatFailedEvent.Failure`.

### CommandFinishedEvent

The type for `ServerConnectionID` has been changed to `*int64` to prevent an `int32` overflow and to differentiate between an ID of `0` and no ID.

### CommandStartedEvent

The type for `ServerConnectionID` has been changed to `*int64` to prevent an `int32` overflow and to differentiate between an ID of `0` and no ID.

### ServerHeartbeatFailedEvent

`DurationNanos` has been removed in favor of `Duration`.

### ServerHeartbeatSucceededEvent

`DurationNanos` has been removed in favor of `Duration`.

## Mongo Package

### Client

#### Connect

> `Client.Connect` has been removed in favor of `mongo.Connect()`. See section on `NewClient` for more details.

Reviewer: Use a consistent format to present methods

```suggestion
`Client.Connect()` has been removed in favor of `mongo.Connect()`. See the section on `NewClient` for more details.
```

Reviewer: applies in multiple places

> The `context.Context` parameter has been removed from `mongo.Connect()` since [deployment connector](https://github.com/mongodb/mongo-go-driver/blob/a76687682f080c9612295b646a00650d00dd16e1/x/mongo/driver/driver.go#L91) doesn’t accept a context, so the context passed to `mongo.Connect` doesn’t actually do anything.

Reviewer: ```suggestion
The `context.Context` parameter has been removed from `mongo.Connect()` because the [deployment connector](https://github.com/mongodb/mongo-go-driver/blob/a76687682f080c9612295b646a00650d00dd16e1/x/mongo/driver/driver.go#L91) doesn’t accept a context, meaning that the context passed to `mongo.Connect()` in previous versions didn't serve a purpose.
```

> The `context.Context` parameter has been removed from `mongo.Connect()` since [deployment connector ](https://github.com/mongodb/mongo-go-driver/blob/a76687682f080c9612295b646a00650d00dd16e1/x/mongo/driver/driver.go#L91) doesn’t accept a context, so the context passed to `mongo.Connect` doesn’t actually do anything.

Reviewer 2: Extra space in [deployment connector ]
```suggestion
The `context.Context` parameter has been removed from `mongo.Connect()` since [deployment connector](https://github.com/mongodb/mongo-go-driver/blob/a76687682f080c9612295b646a00650d00dd16e1/x/mongo/driver/driver.go#L91) doesn’t accept a context, so the context passed to `mongo.Connect` doesn’t actually do anything.
```

#### UseSession\[WithOptions\]

This example shows how to construct a session object from a context, rather than using a context to perform session operations.

```go
// v1

client.UseSession(context.TODO(), func(sctx mongo.SessionContext) error {
  if err := sctx.StartTransaction(options.Transaction()); err != nil {
    return err
  }

  _, err = coll.InsertOne(context.TODO(), bson.D{{"x", 1}})

  return err
})
```

```go
// v2

client.UseSession(context.TODO(), func(ctx context.Context) error {
  sess := mongo.SessionFromContext(ctx)

  if err := sess.StartTransaction(options.Transaction()); err != nil {
    return err
  }

  _, err = coll.InsertOne(context.TODO(), bson.D{{"x", 1}})

  return err
})
```

### Collection

#### Clone

This example shows how to migrate usage of the `collection.Clone` method, which no longer returns an error.

```go
// v1

clonedColl, err := coll.Clone(options.Collection())
if err != nil {
  log.Fatalf("failed to clone collection: %v", err)
}
```

```go
// v2

clonedColl := coll.Clone(options.Collection())
```

#### Distinct

> As of v2, the Distinct collection method will return a struct that can be decoded, similar to `Collection.FindOne`. Instead of iterating through an untyped slice, users can decode same-type data using conventional Go syntax.

Reviewer: ```suggestion
The `Distinct()` collection method returns a struct that can be decoded, similar to `Collection.FindOne`. Instead of iterating through an untyped slice, users can decode same-type data using conventional Go syntax.
```

}

fmt.Printf("car-renting persons: %v\n", people)
```

#### InsertMany

The `documents` parameter in the `Collection.InsertMany` function signature has been changed from an `[]interface{}` type to an `any` type. This API no longer requires users to copy existing slice of documents to an `[]interface{}` slice.

```go
// v1

books := []book{
  {
    Name:   "Don Quixote de la Mancha",
    Author: "Miguel de Cervantes",
  },
  {
    Name:   "Cien años de soledad",
    Author: "Gabriel García Márquez",
  },
  {
    Name:   "Crónica de una muerte anunciada",
    Author: "Gabriel García Márquez",
  },
}

booksi := make([]interface{}, len(books))
for i, book := range books {
  booksi[i] = book
}

_, err = collection.InsertMany(ctx, booksi)
if err != nil {
  log.Fatalf("could not insert Spanish authors: %v", err)
}
```

```go
// v2

books := []book{
  {
    Name:   "Don Quixote de la Mancha",
    Author: "Miguel de Cervantes",
  },
  {
    Name:   "Cien años de soledad",
    Author: "Gabriel García Márquez",
  },
  {
    Name:   "Crónica de una muerte anunciada",
    Author: "Gabriel García Márquez",
  },
}
```

### Database

#### ListCollectionSpecifications

> Updated to return a slice of structs instead of a slice of pointers.

Reviewer: ```suggestion
`ListCollectionSpecifications()` returns a slice of structs instead of a slice of pointers.
```

```

#### ErrWrongIndex

`ErrWrongIndex` has been renamed to the more intuitive `ErrMissingChunk`, which indicates that the number of chunks read from the server is less than expected.

```go
// v1

n, err := source.Read(buf)
if errors.Is(err, gridfs.ErrWrongIndex) {
  // Do something
}
```

```go
// v2

n, err := source.Read(buf)
if errors.Is(err, mongo.ErrMissingChunk) {
  // Do something
}
```

#### SetWriteDeadline

`SetWriteDeadline` methods have been removed from GridFS operations in favor of extending bucket methods to include a `context.Context` argument.

```go
// v1

uploadStream, _ := bucket.OpenUploadStream("filename", uploadOpts)
uploadStream.SetWriteDeadline(time.Now().Add(2*time.Second))
```

```go
// v2

ctx, cancel := context.WithTimeout(context.TODO(), 2*time.Second)
defer cancel()

uploadStream, _ := bucket.OpenUploadStream(ctx, "filename", uploadOpts)
```

Additionally, `Bucket.DeleteContext()`, `Bucket.FindContext()`, `Bucket.DropContext()`, and `Bucket.RenameContext()` have been removed.

### IndexOptionsBuilder

`mongo.IndexOptionsBuilder` has been removed, use the `IndexOptions` type in the `options` package instead.

### IndexView

#### DropAll

> Dropping an index replies with a superset of the following message: `{nIndexesWas: n}`, where n indicates the number of indexes there were prior to removing whichever index(es) were dropped. In the case of `DropAll` this number is always `m - 1`, where m is the total number of indexes. Thus, we can simplify the `DropAll` method by removing the server response.

Reviewer: ```suggestion
Dropping an index replies with a superset of the following message: `{nIndexesWas: n}`, where n indicates the number of indexes there were prior to removing whichever index(es) were dropped. In the case of `DropAll` this number is always `m - 1`, where m is the total number of indexes as you cannot delete the index on the `_id` field. Thus, we can simplify the `DropAll` method by removing the server response.
```

Author: Nice clarification

  log.Fatalf("failed to drop indexes: %v", err)
}

// List the indexes
cur, err := coll.Indexes().List(context.TODO())
if err != nil {
  log.Fatalf("failed to list indexes: %v", err)
}

// Use numDropped
```

#### DropOne

Dropping an index replies with a superset of the following message: `{nIndexesWas: n}`, where n indicates the number of indexes there were prior to removing whichever index(es) were dropped. In the case of `DropOne` this number is always 1 in non-error cases. We can simplify the `DropOne` method by removing the server response.

#### ListSpecifications

Updated to return a slice of structs instead of a slice of pointers. See the [database analogue](#ListCollectionSpecifications) for migration guide.

### NewClient

`NewClient` has been removed, use the `Connect` function in the `mongo` package instead.

```go
client, err := mongo.NewClient(options.Client())
if err != nil {
  log.Fatalf("failed to create client: %v", err)
}

if err := client.Connect(context.TODO()); err != nil {
  log.Fatalf("failed to connect to server: %v", err)
}
```

```go
client, err := mongo.Connect(options.Client())
if err != nil {
  log.Fatalf("failed to connect to server: %v", err)
}

> ```

Reviewer: S: add the v1 and v2 code comments for clarity

  sess, _ := client.StartSession()
  sessions = append(sessions, sess)
}
```

```go
// v2

var sessions []*mongo.Session
for i := 0; i < numSessions; i++ {
  sess, _ := client.StartSession()
  sessions = append(sessions, sess)
}
```

### SingleResult

`SingleResult.DecodeBytes` has been renamed to the more intuitive `SingleResult.Raw`.

### WithSession

This example shows how to update the callback for `mongo.WithSession` to use a `context.Context` implementation, rather than the custom `mongo.SessionContext`.

```go
// v1

mongo.WithSession(context.TODO(),sess,func(sctx mongo.SessionContext) error {
  // Callback
  return nil
})
```

```go
// v2

mongo.WithSession(context.TODO(),sess,func(ctx context.Context) error {
  // Callback
  return nil
})
```

## Options Package

The following fields were marked for internal use only and do not have replacement:

- `ClientOptions.AuthenticateToAnything`
- `FindOptions.OplogReplay`
- `FindOneOptions.OplogReplay`

The following fields were removed because they are no longer supported by the server

- `FindOptions.Snapshot` (4.0)
- `FindOneOptions.Snapshot` (4.0)
- `IndexOptions.Background` (4.2)

### Options

> The Go driver offers users the ability to pass multiple options objects to operations in a last-on-wins algorithm, merging data at a field level:

Reviewer 2: ```suggestion
The Go Driver offers users the ability to pass multiple options objects to operations in a last-on-wins algorithm, merging data at a field level:
```

Using options created with the builder pattern as elements in a slice:

```go
// v1

opts1 := options.Client().SetAppName("appName")
opts2 := options.Client().SetConnectTimeout(math.MaxInt64)

opts := []*options.ClientOptions{opts1, opts2}
_, err := mongo.Connect(opts...)
```

```go
// v2

opts1 := options.Client().SetAppName("appName")
opts2 := options.Client().SetConnectTimeout(math.MaxInt64)

// Option objects are "Listers" in v2, objects that hold a list of setters
opts := []options.Lister[options.ClientOptions]{opts1, opts2}
_, err := mongo.Connect(opts...)
```

#### Creating Options from Builder

Since a builder is just a slice of option setters, users can create options directly from a builder:

```go
// v1

opt := &options.ClientOptions{}
opt.ApplyURI(uri)

return clientOptionAdder{option: opt}
```

```go
// v2

var opts options.ClientOptions
for _, set := range options.Client().ApplyURI(uri).Opts {
  _ = set(&opts)
}

return clientOptionAdder{option: &opts}
```

### FindOneOptions

The following types are not valid for a `findOne` operation and have been removed:

- `BatchSize`
- `CursorType`
- `MaxAwaitTime`
- `NoCursorTimeout`

### Merge\*Options

> All functions that merge options have been removed in favor of a generic solution. See [GORIVER-2696 ](https://jira.mongodb.org/browse/GORIVER-2696)for more information.

Reviewer 2: ```suggestion
All functions that merge options have been removed in favor of a generic solution. See [GODRIVER-2696 ](https://jira.mongodb.org/browse/GODRIVER-2696)for more information.
```

> All functions that merge options have been removed in favor of a generic solution. See [GODRIVER-2696 ](https://jira.mongodb.org/browse/GODRIVER-2696)for more information.

Reviewer 3: ```suggestion
All functions that merge options have been removed in favor of a generic solution. See [GODRIVER-2696](https://jira.mongodb.org/browse/GODRIVER-2696) for more information.
```

- `AggregateOptions.MaxTime` and `AggregateOptions.SetMaxTime`
- `ClientOptions.SocketTimeout` and `ClientOptions.SetSocketTimeout`
- `CountOptions.MaxTime` and `CountOptions.SetMaxTime`
- `DistinctOptions.MaxTime` and `DistinctOptions.SetMaxTime`
- `EstimatedDocumentCountOptions.MaxTime` and `EstimatedDocumentCountOptions.SetMaxTime`
- `FindOptions.MaxTime` and `FindOptions.SetMaxTime`
- `FindOneOptions.MaxTime` and `FindOneOptions.SetMaxTime`
- `FindOneAndReplaceOptions.MaxTime` and `FindOneAndReplaceOptions.SetMaxTime`
- `FindOneAndUpdateOptions.MaxTime` and `FindOneAndUpdateOptions.SetMaxTime`
- `GridFSFindOptions.MaxTime` and `GridFSFindOptions.SetMaxTime`
- `CreateIndexesOptions.MaxTime` and `CreateIndexesOptions.SetMaxTime`
- `DropIndexesOptions.MaxTime` and `DropIndexesOptions.SetMaxTime`
- `ListIndexesOptions.MaxTime` and `ListIndexesOptions.SetMaxTime`
- `SessionOptions.DefaultMaxCommitTime` and `SessionOptions.SetDefaultMaxCommitTime`
- `TransactionOptions.MaxCommitTime` and `TransactionOptions.SetMaxCommitTime`

This example illustrates how to define an operation-level timeout using v2, without loss of generality.

### SessionOptions

> `DefaultReadConcern`, `DefaultReadPreference`, and `DefaultWriteConcern` are all specific to transactions started by the session. Rather than maintain three fields on the Session struct, v2 has combined these options into `DefaultTransactionOptions` which specifies a `TransactionOptions` object.

Reviewer 2: ```suggestion
`DefaultReadConcern`, `DefaultReadPreference`, and `DefaultWriteConcern` are all specific to transactions started by the session. Rather than maintain three fields on the `Session` struct, v2 has combined these options into `DefaultTransactionOptions` which specifies a `TransactionOptions` object.
```

### Options Builder

This example shows how to update usage of `New()` and `Level()` options builder with a `ReadConcern` literal declaration.

```go
// v1

localRC := readconcern.New(readconcern.Level("local"))
```

```go
// v2

localRC := &readconcern.ReadConcern{Level: "local"}
```

### ReadConcern.GetLevel()

The `ReadConcernGetLevel()` method has been removed. Use the `ReadConcern.Level` field to get the level instead.

## Write Concern Package

### WTimeout

The `WTimeout` field has been removed from the `WriteConcern` struct. Instead, users should define a timeout at the operation-level using a context object.

## Bsoncodecs / Bsonoptions Package

`*Codec` structs and `New*Codec` methods have been removed. Additionally, the correlated `bson/bsonoptions` package has been removed, so codecs are not directly configurable using `*CodecOptions` structs in Go Driver 2.0. To configure the encode and decode behavior, use the configuration methods on a `bson.Encoder` or `bson.Decoder`. To configure the encode and decode behavior for a `mongo.Client`, use `options.ClientOptionsBuilder.SetBSONOptions` with `BSONOptions`.

> This example shows how to set ObjectIDAsHex.

Reviewer 2: ```suggestion
This example shows how to set `ObjectIDAsHex`.
```

The `bson.D.String` and `bson.M.String` methods return a relaxed Extended JSON representation of the document.

```go
// v2

d := D{{"a", 1}, {"b", 2}}
fmt.Printf("%s\n", d)
// Output: {"a":{"$numberInt":"1"},"b":{"$numberInt":"2"}}
```

## Bson Package

### DefaultRegistry / NewRegistryBuilder

`DefaultRegistry` has been removed. Using `bson.DefaultRegistry` to either access the default registry behavior or to globally modify the default registry, will be impacted by this change and will need to configure their registry using another mechanism.

The `NewRegistryBuilder` function has been removed along with the `bsoncodec.RegistryBuilder` struct as mentioned above.

### Decoder

#### NewDecoder

The signature of `NewDecoder` has been updated without an error being returned.

#### NewDecoderWithContext

`NewDecoderWithContext` has been removed in favor of using the `SetRegistry` method to set a registry.

Correspondingly, the following methods have been removed:

- `UnmarshalWithRegistry`
- `UnmarshalWithContext`
- `UnmarshalValueWithRegistry`
- `UnmarshalExtJSONWithRegistry`
- `UnmarshalExtJSONWithContext`

> For example, a boolean type can be stored in the database as a BSON boolean, int32, or int64. Given a registry:

Reviewer 2: ```suggestion
For example, a boolean type can be stored in the database as a BSON `Boolean`, `Int32`, or `Int64`. Given a registry:
```

Author: Go native types are not capitalized.

Reviewer 3: I regard "boolean", "in32", "int64", or "true" are general concepts, that don't need to be denoted.

Reviewer 2: @prestonvasquez isn't this specifically talking about the BSON types rather than the native types? This suggestion was based on the code sample below which has `bsontype.Boolean`, `bsontype.Int32`, and `bsontype.Int64`

Reviewer 2: Ok, I'll defer to your judgement here. I noticed that it was capitalized in the code sample, which is why I suggested it, but if it doesn't make sense, let's keep it as-is. Feel free to resolve this one :)

```go
type lenientBool bool

lenientBoolType := reflect.TypeOf(lenientBool(true))

lenientBoolDecoder := func(
	dc bsoncodec.DecodeContext,
	vr bsonrw.ValueReader,
	val reflect.Value,
) error {
	// All decoder implementations should check that val is valid, settable,
	// and is of the correct kind before proceeding.
	if !val.IsValid() || !val.CanSet() || val.Type() != lenientBoolType {
		return bsoncodec.ValueDecoderError{
			Name:     "lenientBoolDecoder",
			Types:    []reflect.Type{lenientBoolType},
			Received: val,
		}
	}

	var result bool
	switch vr.Type() {
	case bsontype.Boolean:
		b, err := vr.ReadBoolean()
		if err != nil {
			return err
		}
		result = b
	case bsontype.Int32:
		i32, err := vr.ReadInt32()
		if err != nil {
			return err
		}
		result = i32 != 0
	case bsontype.Int64:
		i64, err := vr.ReadInt64()
		if err != nil {
			return err
		}
		result = i64 != 0
	default:
		return fmt.Errorf(
			"received invalid BSON type to decode into lenientBool: %s",
			vr.Type())
	}

	val.SetBool(result)
	return nil
}

// Create the registry
reg := bson.NewRegistry()
reg.RegisterTypeDecoder(
	lenientBoolType,
	bsoncodec.ValueDecoderFunc(lenientBoolDecoder))
```

> For our custom decoder with such a registry, BSON int32 or int64 values are considered "true" if they are non-zero.

Reviewer 2: ```suggestion
For our custom decoder with such a registry, BSON `Int32` or `Int64` values are considered `true` if they are non-zero.
```

Reviewer 2: We can keep the int32 and int64 lowercase as discussed in the previous thread.
But I'm guessing we should remove the quotation around "true" though. It's probably best to format it as `code` (based off of the convention I'm noticing in the rest of the doc when it comes to values, for example "to differentiate between an ID of `0` and no ID").

#### ObjectIDAsHexString

`Decoder.ObjectIDAsHexString` method enables decoding a BSON Object ID as a hexadecimal string. Otherwise, the decoder returns an error by default instead of decoding as the UTF-8 representation of the raw Object ID bytes, which results in a garbled and unusable string.

### Encoder

#### NewEncoder

The signature of `NewEncoder` has been updated without an error being returned.

#### NewEncoderWithContext

`NewEncoderWithContext` has been removed in favor of using the `SetRegistry` method to set a registry.

Correspondingly, the following methods have been removed:

- `MarshalWithRegistry`
- `MarshalWithContext`
- `MarshalAppend`
- `MarshalAppendWithRegistry`
- `MarshalAppendWithContext`
- `MarshalValueWithRegistry`
- `MarshalValueWithContext`
- `MarshalValueAppendWithRegistry`
- `MarshalValueAppendWithContext`
- `MarshalExtJSONWithRegistry`
- `MarshalExtJSONWithContext`
- `MarshalExtJSONAppendWithRegistry`
- `MarshalExtJSONAppendWithContext`

> Here is an example of a registry that multiplies the input value by -1 when encoding for a `negatedInt.`

Reviewer 2: ```suggestion
Here is an example of a registry that multiplies the input value by -1 when encoding for a `negatedInt`.
```

if err != nil {
	panic(err)
}
ec := bsoncodec.EncodeContext{Registry: reg}
enc, err := bson.NewEncoderWithContext(ec, vw)
if err != nil {
	panic(err)
}
err = enc.Encode(bson.D{{"negatedInt", negatedInt(1)}})
if err != nil {
	panic(err)
}
fmt.Println(bson.Raw(buf.Bytes()).String())
// Output: {"negatedint": {"$numberInt":"-1"}}
```

```go
// v2
// Use SetRegistry

buf := new(bytes.Buffer)
vw := bson.NewDocumentWriter(buf)
enc := bson.NewEncoder(vw)
enc.SetRegistry(reg)
err := enc.Encode(bson.D{{"negatedInt", negatedInt(1)}})
if err != nil {
	panic(err)
}
fmt.Println(bson.Raw(buf.Bytes()).String())
// Output: {"negatedint": {"$numberInt":"-1"}}
```

#### SetContext

The `SetContext` method has been removed in favor of using `SetRegistry` to set the registry of an encoder.

#### SetRegistry

The signature of `SetRegistry` has been updated without an error being returned.

#### Reset

The signature of `Reset` has been updated without an error being returned.

### RawArray Type

A new `RawArray` type has been added to the `bson` package as a primitive type to ​​represent a BSON array. Correspondingly, `RawValue.Array()` returns a `RawArray` instead of `Raw`.

### ValueMarshaler

> The `MarshalBSONValue` method of the `ValueMarshaler` interface is only required to return a byte type value representing the BSON type to avoid importing the Go driver package.

Reviewer 2: What's the name of the package that no longer needs to be imported?

Reviewer 2: @qingyang-hu?

Reviewer 3: I've updated at https://github.com/mongodb/mongo-go-driver/pull/1766/commits/60af8e342b6d20a0166e05c3cc06e81c79d8b6a6

```

```go
// v2

filter := bson.D{{"age", bson.D{{"$gt", 25}}}}

res := coll.Distinct(context.TODO(), "name", filter)
if err := res.Err(); err != nil {
  log.Fatalf("failed to get distinct result: %v", err)
}

var people []string
if err := res.Decode(&people); err != nil {
  log.Fatal("failed to decode distinct result: %v", err)
}

fmt.Printf("car-renting persons: %v\n", people)
```

If the data returned is not same-type (i.e. `name` is not always a string) a user can iterate through the result directly as a `bson.RawArray` type:

```go
// v2

filter := bson.D{{"age", bson.D{{"$gt", 25}}}}
distinctOpts := options.Distinct().SetMaxTime(2 * time.Second)

res := coll.Distinct(context.TODO(), "name", filter, distinctOpts)
if err := res.Err(); err != nil {
  log.Fatalf("failed to get distinct result: %v", err)
}

rawArr, err := res.Raw()
if err != nil {
  log.Fatalf("failed to get raw data: %v", err)
}

values, err := rawArr.Values()
if err != nil {
  log.Fatalf("failed to get values: %v", err)
}

people := make([]string, 0, len(rawArr))
for _, value := range values {
  people = append(people, value.String())
}

fmt.Printf("car-renting persons: %v\n", people)
```

> ### InsertMany

Reviewer 3: ```suggestion
#### InsertMany
```

The `documents` parameter in the `Collection.InsertMany` function signature has been changed from an `[]interface{}` type to an `any` type. This API no longer requires users to copy existing slice of documents to an `[]interface{}` slice.

```go
// v1

books := []book{
  {
    Name:   "Don Quixote de la Mancha",
    Author: "Miguel de Cervantes",
  },
  {
    Name:   "Cien años de soledad",
    Author: "Gabriel García Márquez",
  },
  {
    Name:   "Crónica de una muerte anunciada",
    Author: "Gabriel García Márquez",
  },
}

booksi := make([]interface{}, len(books))
for i, book := range books {
  booksi[i] = book
}

_, err = collection.InsertMany(ctx, booksi)
if err != nil {
  log.Fatalf("could not insert Spanish authors: %v", err)
}
```

> ````go

Reviewer 3: ```suggestion
```go
```
